src/libs/actions/Transaction.ts

```diff
@@ -572,6 +584,139 @@ function getAllTransactions() {
     return Object.keys(allTransactions ?? {}).length;
 }
 
+function changeTransactionsReport(transactionIDs: string[], reportID: string, transactionIDToReportActionAndThreadData: Record<string, string>) {
+    const optimisticData: OnyxUpdate[] = [];
+    const failureData: OnyxUpdate[] = [];
+    const successData: OnyxUpdate[] = [];
+
+    const newPolicyID = allReports?.[`${ONYXKEYS.COLLECTION.REPORT}${reportID}`]?.policyID;
+
+    // 1. Optimistically change the reportID on the passed transactions
+    const optimisticDataTransactions: OnyxUpdate[] = transactionIDs.map((transactionID) => ({
+        onyxMethod: Onyx.METHOD.MERGE,
+        key: `${ONYXKEYS.COLLECTION.TRANSACTION}${transactionID}`,
+        value: {
+            reportID,
+        },
+    }));
+
+    const failureDataTransactions: OnyxUpdate[] = transactionIDs.map((transactionID) => {
+        const transaction = allTransactions?.[transactionID];
+        return {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.TRANSACTION}${transactionID}`,
+            value: {
+                reportID: transaction?.reportID,
+            },
+        };
+    });
+
+    optimisticData.push(...optimisticDataTransactions);
+    failureData.push(...failureDataTransactions);
+
+    // 2. Optimistically update the IOU report action reportID
+    const optimisticDataReportActions: OnyxUpdate[] = Object.values(transactionIDToReportActionAndThreadData).map((reportActionID) => ({
+        onyxMethod: Onyx.METHOD.MERGE,
+        key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${reportActionID}`,
+        value: {
+            reportID,
+        },
+    }));
+
+    const failureDataReportActions: OnyxUpdate[] = Object.entries(transactionIDToReportActionAndThreadData).map(([transactionID, reportActionID]) => {
+        const transaction = allTransactions?.[transactionID];
+        return {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${reportActionID}`,
+            value: {
+                reportID: transaction?.reportID,
+            },
+        };
+    });
+
+    optimisticData.push(...optimisticDataReportActions);
+    failureData.push(...failureDataReportActions);
+
+    // 3. Optimistically update the transaction thread and all threads in the transaction thread
+    const optimisticDataThreadReports: OnyxUpdate[] = Object.values(transactionIDToReportActionAndThreadData).map((threadReportID) => ({
```

Comment:
Here again, `transactionIDToReportActionAndThreadData` should only stores the `transactionThreadReportID` if we're optimistically creating the transactionThread